### PR TITLE
Android marker cluster

### PR DESCRIPTION
**Pull request by bosung90**
1. Android marker clustering using Google Maps Android Utils
2. When marker image is provided directly as image prop, cache the image so new BitmapDescriptor doesn't have to be created for every single marker.
3. Tap on cluster auto zooms into the cluster
4. Adds hole for polygon

### CODE REVIEW

**Comment by Jazz747** [comment]
Hi, this looks promising, how can I install react-native-maps with your changes, I want to test this feature?

**Comment by bosung90** [comment]
@Jazz747 You can target your npm to my commit directly.
https://docs.npmjs.com/files/package.json#git-urls-as-dependencies

**Comment by henrikra** [comment]
@bosung90 Provide screenshot to the description of this pr? :D

**Comment by nehvaleem** [comment]
Sounds awesome! Can't wait for this PR to be merged!

File: lib/android/src/main/java/com/airbnb/android/react/maps/AirMapMarker.java [hunk 1]
```diff
@@ -216,27 +199,79 @@ public void setCalloutAnchor(double x, double y) {
...
             iconBitmapDescriptor = null;
+            iconBitmap = null;
+            update();
+            return;
+        }
+
+
+        try {
+            while(isCacheAdded.contains(uri))
+                Thread.sleep(100);
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/AirMapMarker.java]: Why is this necessary?

File: lib/android/src/main/java/com/airbnb/android/react/maps/AirMapPolygon.java [hunk 2]
```diff
@@ -41,6 +44,41 @@ public void setCoordinates(ReadableArray coordinates) {
         }
     }
 
+    public void setHoles(ReadableArray holes) {
+        // Convert ReadableArray<ReadableArray<ReadableMap>> into List<List<LatLng>>
+        int holesSize = holes.size();
+        this.holes = new ArrayList(holesSize);
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/AirMapPolygon.java]: use `new ArrayList<>(holesSize);` instead

File: lib/android/src/main/java/com/airbnb/android/react/maps/AirMapPolygon.java [hunk 3]
```diff
@@ -41,6 +44,41 @@ public void setCoordinates(ReadableArray coordinates) {
...
+                    ReadableMap coordinate0 = hole.getMap(0);
+                    double lat0 = coordinate0.getDouble("latitude");
+                    double lon0 = coordinate0.getDouble("longitude");
+
+                    ReadableMap coordinate1 = hole.getMap(hole.size()-1);
+                    double lat1 = coordinate1.getDouble("latitude");
+                    double lon1 = coordinate1.getDouble("longitude");
+
+                    if(lat0 != lat1 || lon0 != lon1) {
+                        Log.e(TAG, "First and Last coordinates of the holes must be same");
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/AirMapPolygon.java]: Why only do this for debug builds?

File: lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java [hunk 4]
```diff
@@ -88,6 +99,9 @@
     private final ThemedReactContext context;
     private final EventDispatcher eventDispatcher;
 
+    private ClusterManager<AirMapMarker> mClusterManager;
+    private ClusterMarkerIconGenerator mClusterIconGenerator = null;
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java]: we don't use `mPrefix`

File: lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java [hunk 5]
```diff
@@ -170,6 +184,68 @@ public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX,
         });
 
         eventDispatcher = reactContext.getNativeModule(UIManagerModule.class).getEventDispatcher();
+
+        mClusterIconGenerator = new ClusterMarkerIconGenerator(reactContext);
+
+    }
+
+    private class MarkerRenderer extends DefaultClusterRenderer<AirMapMarker> {
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java]: move this to a separate class

File: lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java [hunk 6]
```diff
@@ -180,38 +256,72 @@ public void onMapReady(final GoogleMap map) {
...
+//                    event = makeClickEventData(marker.getPosition());
+//                    event.putString("action", "marker-press");
+//                    event.putString("id", airMapMarker.getIdentifier());
+//                    event.putInt("count", 1);
+//                    event.putDouble("zoom", map.getCameraPosition().zoom);
+//                    manager.pushEvent(context, view, "onPress", event);
+//                    return true;
+//                }
+//            }
+//        });
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java]: just delete the code instead of commenting it out

File: lib/android/src/main/java/com/airbnb/android/react/maps/BitmapDescriptorContainer.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package com.airbnb.android.react.maps;
+
+import android.graphics.Bitmap;
+
+import com.google.android.gms.maps.model.BitmapDescriptor;
+import com.google.android.gms.maps.model.BitmapDescriptorFactory;
+
+/**
+ * Created by Eric Kim on 2017-04-10.
+ */
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/BitmapDescriptorContainer.java]: remove

File: lib/android/src/main/java/com/airbnb/android/react/maps/BitmapDescriptorContainer.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+public class BitmapDescriptorContainer
+{
+    private int mBitmapByteCount;
+    public BitmapDescriptor mBitmapDescriptor;
+    public Bitmap mBitmap;
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/BitmapDescriptorContainer.java]: make fields final

File: lib/android/src/main/java/com/airbnb/android/react/maps/BitmapDescriptorContainer.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+    public BitmapDescriptorContainer(Bitmap bitmap)
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/BitmapDescriptorContainer.java]: open curly braces should go on the same line as the method name

File: lib/android/src/main/java/com/airbnb/android/react/maps/ClusterMarkerIconGenerator.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+import android.graphics.Rect;
+import android.graphics.drawable.Drawable;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.widget.RelativeLayout;
+import android.widget.TextView;
+
+/**
+ * Created by Eric Kim on 2017-05-08.
+ */
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/ClusterMarkerIconGenerator.java]: remove

File: lib/android/src/main/java/com/airbnb/android/react/maps/ClusterMarkerIconGenerator.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+public class ClusterMarkerIconGenerator {
+    private final Context mContext;
+
+    private RelativeLayout mContainer;
+    private TextView mTextView;
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/ClusterMarkerIconGenerator.java]: make them final, remove m prefix

File: lib/android/src/main/java/com/airbnb/android/react/maps/LruCacheManager.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+import android.graphics.Bitmap;
+import android.util.LruCache;
+
+import com.google.android.gms.maps.model.BitmapDescriptor;
+import com.google.android.gms.maps.model.BitmapDescriptorFactory;
+
+/**
+ * Created by Eric Kim on 2017-04-07.
+ */
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/LruCacheManager.java]: remove

File: lib/android/src/main/java/com/airbnb/android/react/maps/LruCacheManager.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+public class LruCacheManager {
+    private static final String TAG = LruCacheManager.class.getName();
+
+    private LruCache<String, BitmapDescriptorContainer> mMemoryCache;
+    private static final LruCacheManager INSTANCE = new LruCacheManager();
```
Comment (felipecsl) [inline lib/android/src/main/java/com/airbnb/android/react/maps/LruCacheManager.java]: if you want to make it a singleton, make the constructor private

File: lib/android/src/main/java/com/airbnb/android/react/maps/AirMapMarker.java (hunk 1 above)
Comment (ericapply) [inline lib/android/src/main/java/com/airbnb/android/react/maps/AirMapMarker.java]: If other thread is already loading a same image, wait for it to finish

File: lib/android/src/main/java/com/airbnb/android/react/maps/AirMapPolygon.java (hunk 3 above)
Comment (ericapply) [inline lib/android/src/main/java/com/airbnb/android/react/maps/AirMapPolygon.java]: Because the hole will fail silently. No need to do these checks if we can avoid it.

**Comment by RWOverdijk** [comment]
This is very cool. I'd like something like this. Is this possible for iOS? If so what is needed / how can we help?

**Comment by ericapply** [comment]
@RWOverdijk I already did it for ios google maps but its not ready for PR
https://github.com/ericapply/react-native-maps/tree/ios-clustering

**Comment by nehvaleem** [comment]
@RWOverdijk - I belive there is already an open PR that addresses this same case for iOS :)

@bosung90 - are you gonna try to work further with this PR or you need some assistance? It's been some time :)

**Comment by bosung90** [comment]
@nehvaleem If you can assist it would be great. Thank you!!

**Comment by RWOverdijk** [comment]
In that case, awesome! this feature would make the user experience a lot better.

**Comment by xernaj** [comment]
It looks like there's an older PR open for "Adds hole for polygon" for Android  https://github.com/airbnb/react-native-maps/pull/1271.

Hopefully one of them goes in so the feature set for holes is the same as ios.

**Comment by oskaryil** [comment]
Any update on when this goes into upstream?

**Comment by venits** [comment]
You can try this. It works for me :) https://github.com/venits/react-native-map-clustering
It is the simplest solution right now.

**Comment by gvlekke** [comment]
any news on this?

**Comment by cinder92** [comment]
any updatE?

**Comment by rborn** [comment]
This is a very old PR, with unresolved code review and many conflicts so I'll close it.
@cinder92 I used a similar approach with https://github.com/venits/react-native-map-clustering and it works well.

**Comment by NooruddinLakhani** [comment]
> You can try this. It works for me :) https://github.com/venits/react-native-map-clustering
> It is the simplest solution right now.

This is not using Google Clustering, What would be the solution if we have to use google clustering for react-native both for IOS and Android using react-native-maps.
